src/Storages/ObjectStorage/StorageObjectStorageSource.cpp

```diff
@@ -0,0 +1,624 @@
+    read_settings.enable_filesystem_cache = false;
+    /// FIXME: Changing this setting to default value breaks something around parquet reading
+    read_settings.remote_read_min_bytes_for_seek = read_settings.remote_fs_buffer_size;
+
+    const bool object_too_small = object_size <= 2 * getContext()->getSettings().max_download_buffer_size;
+    const bool use_prefetch = object_too_small && read_settings.remote_fs_method == RemoteFSReadMethod::threadpool;
+    read_settings.remote_fs_method = use_prefetch ? RemoteFSReadMethod::threadpool : RemoteFSReadMethod::read;
+    /// User's object may change, don't cache it.
+    read_settings.use_page_cache_for_disks_without_file_cache = false;
+
+    // Create a read buffer that will prefetch the first ~1 MB of the file.
+    // When reading lots of tiny files, this prefetching almost doubles the throughput.
+    // For bigger files, parallel reading is more useful.
+    if (use_prefetch)
+    {
+        LOG_TRACE(log, "Downloading object of size {} with initial prefetch", object_size);
+
+        auto async_reader = object_storage->readObjects(
+            StoredObjects{StoredObject{key, /* local_path */ "", object_size}}, read_settings);
+
+        async_reader->setReadUntilEnd();
+        if (read_settings.remote_fs_prefetch)
+            async_reader->prefetch(DEFAULT_PREFETCH_PRIORITY);
+
+        return async_reader;
+    }
+    else
+    {
+        /// FIXME: this is inconsistent that readObject always reads synchronously ignoring read_method setting.
+        return object_storage->readObject(StoredObject(key), read_settings);
+    }
+}
+
+StorageObjectStorageSource::IIterator::IIterator(const std::string & logger_name_)
+    : logger(getLogger(logger_name_))
+{
+}
+
+StorageObjectStorage::ObjectInfoPtr StorageObjectStorageSource::IIterator::next(size_t processor)
+{
+    auto object_info = nextImpl(processor);
+
+    if (object_info)
+    {
+        LOG_TEST(logger, "Next key: {}", object_info->relative_path);
+    }
+
+    return object_info;
+}
+
+StorageObjectStorageSource::GlobIterator::GlobIterator(
+    ObjectStoragePtr object_storage_,
+    ConfigurationPtr configuration_,
+    const ActionsDAG::Node * predicate,
+    const NamesAndTypesList & virtual_columns_,
+    ContextPtr context_,
+    ObjectInfos * read_keys_,
+    size_t list_object_keys_size,
+    bool throw_on_zero_files_match_,
+    std::function<void(FileProgress)> file_progress_callback_)
+    : IIterator("GlobIterator")
+    , WithContext(context_)
+    , object_storage(object_storage_)
+    , configuration(configuration_)
+    , virtual_columns(virtual_columns_)
+    , throw_on_zero_files_match(throw_on_zero_files_match_)
+    , read_keys(read_keys_)
+    , file_progress_callback(file_progress_callback_)
+{
+    if (configuration->isNamespaceWithGlobs())
+    {
+        throw Exception(ErrorCodes::BAD_ARGUMENTS, "Expression can not have wildcards inside namespace name");
```

Comment:
But we already have the same check in `createFileIterator`. Or it's not enough?

Reply:
If we do create those iterators outside of `createFileIterator` method, then it makes sense to have in both places I'd say. Also accounting that in `createFileIterator` we create glob iterator only if `isPathWithGlobs`, so we could create keys iterator, for example, and have globs in namespace, so to avoid this case we check for globs in namespace in `getFileIterator`.